### ISSUE TITLE
PT2 inference slowdown after updating Huggingface pin on CI

### ISSUE THREAD

**Issue by desertfire** [pytorch/pytorch#133462]
From https://github.com/pytorch/pytorch/pull/133065#issuecomment-2288701447 . Basically, there was noticeable performance drop on the inference side after bumping up the HF pin, [dashboard](https://hud.pytorch.org/benchmark/compilers?dashboard=torchinductor&startTime=Wed%2C%2007%20Aug%202024%2013%3A09%3A28%20GMT&stopTime=Wed%2C%2014%20Aug%202024%2013%3A09%3A28%20GMT&granularity=hour&suite=torchbench&mode=inference&dtype=bfloat16&deviceName=cuda%20(a100)&lBranch=gh/desertfire/449/head&lCommit=445863995081ff442d69a567431f056c2095f2aa&rBranch=main&rCommit=c8275e25a79903daac3aa5ed4ad8fd3132ca6adb)

**Comment by eellison** [comment]
cc @anijain2305, @yanboliang @mlazos , `BartForCausalLM` goes from 12.9290 → 17.0171 execution time. anytime we've seen a slowdown like this it has been from dynamo changes. have any bandwidth to take a look ?

**Comment by eellison** [comment]
Okay, looked into it a bit more - at least for `BertForMaskedLM`, no TORCH_LOGS="graph_breaks,recompiles" so i dont think it it's a dynamo issue. 

Here is output code before and after: https://www.diffchecker.com/kEW9m4Ie/

Two things:

- HF is explicitly invoking `SDPA` now, so it makes sense that it inductor slowdown is lower.
- We're going from invoking `_scaled_dot_product_flash_attention` to `_scaled_dot_product_efficient_attention` which accounts for absolute execution time going down. 
```        buf8 = torch.ops.aten._scaled_dot_product_flash_attention.default(reinterpret_tensor(buf5, (16, 12, 512, 64), (393216, 64, 768, 1), 0), reinterpret_tensor(buf6, (16, 12, 512, 64), (393216, 64, 768, 1), 0), reinterpret_tensor(buf7, (16, 12, 512, 64), (393216, 64, 768, 1), 0), scale=0.125)
```
vs 
```
buf9 = torch.ops.aten._scaled_dot_product_efficient_attention.default(reinterpret_tensor(buf5, (16, 12, 512, 64), (393216, 64, 768, 1), 0), reinterpret_tensor(buf6, (16, 12, 512, 64), (393216, 64, 768, 1), 0), reinterpret_tensor(buf7, (16, 12, 512, 64), (393216, 64, 768, 1), 0), buf8, False)
```

cc @drisspg, possible to spot check why we're not hitting the FA path now ?

**Comment by drisspg** [comment]
@eellison There hasnt been any new changes that would have made something working with Flash suddenly not, so that is pretty strange

**Comment by eellison** [comment]
@drisspg this would be from model level changes.. I see additional input `buf8`. Maybe bias is the reason we're not hitting it ?

**Comment by drisspg** [comment]
@eellison yeah if they are doing MaskedLM I imagine that they are now providing a mask, where maybe they weren't in the past?

**Comment by eellison** [comment]
I think the issue is we're not constant folding away the all 0 bias. And maybe part of the reason is for some reason we're doing this instead of 0:

```
    %scalar_tensor : [num_users=1] = call_function[target=torch.ops.aten.scalar_tensor.default](args = (-3.3895313892515355e+38,), kwargs = {dtype: torch.bfloat16, layout: torch.strided, device: cuda:0})
```

After constant folding it away to 0, may also need to replace the explicit `_scaled_dot_product_efficient_attention` invocation with `_scaled_dot_product_flash_attention`

**Comment by eellison** [comment]
To fix:

We're bailing on replacing with zeros because the expand causes non contiguous tensor. Need to add a full_strided variant, and remove the remove the is_contiguous check here, see: https://github.com/pytorch/pytorch/blob/2e8c1be947896c443c9461f98f2a47062cd7d7c4/torch/_inductor/fx_passes/joint_graph.py#L358-L366

Also need to remove `constant_data_ptr_count` check [here](https://github.com/pytorch/pytorch/blob/2e8c1be947896c443c9461f98f2a47062cd7d7c4/torch/_inductor/fx_passes/joint_graph.py#L355). ran into error at one point with this, but i think we could just check if they are fx graph outputs.

- Might need to find way to replace the traced out efficient_attention if we've discovered the bias is empty.

**Comment by eellison** [comment]
We know what fix is here but i'm not actively working on it, so un-assigning myself.
